### ISSUE TITLE
use pygit2 in the scm package

### ISSUE THREAD

**Issue by vladak** [oracle/opengrok#2852]
PR #2851 added pygit2 for testing. It might be well used in `opengrok_tools.scm.git` as well.

**Comment by tulinkry** [comment]
I wonder why you chose pygit2 (which even requires some brew install as I saw in your PR) over others:

https://github.com/gitpython-developers/GitPython
https://pypi.org/project/python-git/

**Comment by vladak** [comment]
No particular reason besides pretty clean API - at least for what I needed.

**Comment by tulinkry** [comment]
As I understood the pygit2 uses native api calls and not "exec" style like the others. libgit2 is installed on unix systems? (as you only added it to the osx builds)

**Comment by tulinkry** [comment]
Cause suddenly it does not work out of the box on osx. I have to install libgit2 manually and it is not mentioned anywhere that we have this hidden dependency for development.

**Comment by tulinkry** [comment]
when we used the other packages (with the "exec" style) I'd say you dont need anything than python

**Comment by vladak** [comment]
Good point, I should have added a note to the Readme as I experienced the
same.

Dne po 8. 7. 2019 10:35 uživatel Kryštof Tulinger <notifications@github.com>
napsal:

**Comment by tulinkry** [comment]
Is speed of the calls the concern here? Cause I'd prefer not to install any external dependency - as you plan to use it not only for testing but for the git repo.

**Comment by vladak** [comment]
No, pygit2 looked attractive because of the clarity of the API. You just go `init_repository()`, `clone_repository()` and boom, that's it. While with executing the porcelain commands, one has to care about return codes, error string matching etc.

You are probably right that with this non-implicit dependency this is awkward to use. I will keep this as is for now and document it in the development section of the README.

**Comment by vladak** [comment]
But still, the dependency should be investigates in more detail - how come this works out of the box on (some) Linux distros but not macOS. I would have thought that `pip install` should handle this.

**Comment by vladak** [comment]
I just got bitten by this again, trying to run `python setup.py test` on fresh Solaris install. In hindsight, choosing pygit2 was wrong. We need something that does not require compilation.

**Comment by vladak** [comment]
https://pypi.org/project/python-git/ has very limited API, it lacks clone/init, seems to be not very mature in general.

**Comment by vladak** [comment]
The downside of using Git Python module is that command timeouts would have to implemented differently. Now it is provided by the `Command` class automatically (modulo bugs like #3022).

**Comment by vladak** [comment]
GitPython should be excluded as well because the README says:
> For the time to come I am happy to continue maintaining GitPython, remaining hopeful that one day it won't be needed anymore.

and

> This project is in maintenance mode, which means that
>
>    …there will be no feature development, unless these are contributed
>    …there will be no bug fixes, unless they are relevant to the safety of users, or contributed
>    …issues will be responded to with waiting times of up to a month
>
> The project is open to contributions of all kinds, as well as new maintainers.

which pretty much seems like active discouragement from using it.

**Comment by vladak** [comment]
Looking at python-git, it requires `git` executable so this is not going to work either.

**Comment by vladak** [comment]
Gittle/Dulwich might be workable (from https://stackoverflow.com/questions/5350518/is-there-implementation-of-git-in-pure-python) except Gittle has not been updated for couple of years.
